p2p/nodestate/nodestate.go

```diff
@@ -0,0 +1,966 @@
+
+// decodeNode decodes a node database entry and adds it to the node set if successful
+func (ns *NodeStateMachine) decodeNode(id enode.ID, data []byte) {
+	var enc nodeInfoEnc
+	if err := rlp.DecodeBytes(data, &enc); err != nil {
+		log.Error("Failed to decode node info", "id", id, "error", err)
+		return
+	}
+	n, _ := enode.New(dummyIdentity(id), &enc.Enr)
+	node := ns.newNode(n)
+	node.db = true
+
+	if int(enc.Mapping) >= len(ns.mappings) {
+		log.Error("Unknown scheme", "id", id, "index", enc.Mapping, "len", len(ns.mappings))
+		return
+	}
+	encMapping := ns.mappings[int(enc.Mapping)]
+	if len(enc.Fields) != len(encMapping.Fields) {
+		log.Error("Invalid node field count", "id", id, "stored", len(enc.Fields), "mapping", len(encMapping.Fields))
+		return
+	}
+	// convertMask converts a old format state/mask to the latest version.
+	convertMask := func(schemeID int, scheme []string, state bitMask) (bitMask, bool) {
+		if schemeID == ns.currentMapping {
+			return state, true // Nothing need to be changed
+		}
+		var converted bitMask
+		for i, name := range scheme {
+			if (state & (bitMask(1) << i)) != 0 {
+				if index, ok := ns.stateNameMap[name]; ok {
+					converted |= bitMask(1) << index
+				} else {
+					log.Error("Unknown state flag", "name", name)
+					return bitMask(0), false // unknown flag
+				}
+			}
+		}
+		return converted, true
+	}
+	// Resolve persisted node fields
+	for i, encField := range enc.Fields {
+		if len(encField) == 0 {
+			continue
+		}
+		index := i
+		if int(enc.Mapping) != ns.currentMapping {
+			name := encMapping.Fields[i]
+			var ok bool
+			if index, ok = ns.fieldNameMap[name]; !ok {
+				log.Error("Unknown node field", "id", id, "field name", name)
+				return
+			}
+		}
+		if decode := ns.fields[index].decode; decode != nil {
+			if field, err := decode(encField); err == nil {
+				node.fields[index] = field
+			} else {
+				log.Error("Failed to decode node field", "id", id, "field name", ns.fields[index].name, "error", err)
+				return
+			}
+		} else {
+			log.Error("Cannot decode node field", "id", id, "field name", ns.fields[index].name)
+			return
+		}
+	}
+	// Resolve node state
+	state, success := convertMask(int(enc.Mapping), encMapping.States, enc.State)
+	if !success {
+		return
+	}
+	// It's a compatible node record, add it to set.
+	ns.nodes[id] = node
+	node.state = state
+	fields := make([]interface{}, len(node.fields))
+	copy(fields, node.fields)
+	ns.offlineCallbackList = append(ns.offlineCallbackList, offlineCallback{node.node, state, fields})
+	log.Debug("Loaded node state", "id", id, "state", Flags{mask: enc.State, setup: ns.setup})
+}
+
+// saveNode saves the given node info to the database
+func (ns *NodeStateMachine) saveNode(id enode.ID, node *nodeInfo) error {
+	if ns.db == nil {
+		return nil
+	}
+
+	storedState := node.state & ns.saveFlags
+	for _, t := range node.timeouts {
+		storedState &= ^t.mask
+	}
+	if storedState == 0 {
```

Comment:
I think in this case, these timeouts are applied in advance. So we should get rid of them.
Otherwise it might be buggy this the following example:

- Addtimeout(node, flags)
- Persist(node)
- UpdateState(node, flags')
- Timeout, the re-added flags are removed

Reply:
`Persist` does not change the node state so the flags with the active timeouts are not actually removed and therefore no inconsistency is created here. These flags will stay active until the originally planned timeout. Here we are just calculating the state of the stored version which only includes persistent flags with no timeout.